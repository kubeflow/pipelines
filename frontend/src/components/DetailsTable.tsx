/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { stylesheet } from 'typestyle';
import { color, spacing, commonCss } from '../Css';
import { KeyValue } from '../lib/StaticGraphParser';
import Editor from './Editor';
import 'brace';
import 'brace/ext/language_tools';
import 'brace/mode/json';
import 'brace/theme/github';

export const css = stylesheet({
  key: {
    color: color.strong,
    flex: '0 0 50%',
    fontWeight: 'bold',
    maxWidth: 300,
  },
  row: {
    borderBottom: `1px solid ${color.divider}`,
    display: 'flex',
    padding: `${spacing.units(-5)}px ${spacing.units(-6)}px`,
  },
  valueJson: {
    flexGrow: 1,
  },
  valueText: {
    // flexGrow expands value text to full width.
    flexGrow: 1,
    // For current use-cases, value text shouldn't be very long. It will be not readable when it's long.
    // Therefore, it's easier we just show it completely in the UX.
    overflow: 'hidden',
    // Sometimes, urls will be unbreakable for a long string. overflow-wrap: break-word
    // allows breaking an url at middle of a word so it will not overflow and get hidden.
    overflowWrap: 'break-word',
  },
});

export interface ValueComponentProps<T> {
  value?: string | T;
  [key: string]: any;
}

interface DetailsTableProps<T> {
  fields: Array<KeyValue<string | T>>;
  title?: string;
  valueComponent?: React.FC<ValueComponentProps<T>>;
  valueComponentProps?: { [key: string]: any };
}

function isString(x: any): x is string {
  return typeof x === 'string';
}

const DetailsTable = <T extends {}>(props: DetailsTableProps<T>) => {
  const { fields, title, valueComponent: ValueComponent, valueComponentProps } = props;
  return (
    <React.Fragment>
      {!!title && <div className={commonCss.header}>{title}</div>}
      <div>
        {fields.map((f, i) => {
          const [key, value] = f;

          // only try to parse json if value is a string
          if (isString(value)) {
            try {
              const parsedJson = JSON.parse(value);
              // Nulls, booleans, strings, and numbers can all be parsed as JSON, but we don't care
              // about rendering. Note that `typeOf null` returns 'object'
              if (parsedJson === null || typeof parsedJson !== 'object') {
                throw new Error(
                  'Parsed JSON was neither an array nor an object. Using default renderer',
                );
              }
              return (
                <div key={i} className={css.row}>
                  <span className={css.key}>{key}</span>
                  <Editor
                    width='100%'
                    minLines={3}
                    maxLines={20}
                    mode='json'
                    theme='github'
                    highlightActiveLine={true}
                    showGutter={true}
                    readOnly={true}
                    value={JSON.stringify(parsedJson, null, 2) || ''}
                  />
                </div>
              );
            } catch (err) {
              // do nothing
            }
          }
          // If a ValueComponent and a value is provided, render the value with
          // the ValueComponent. Otherwise render the value as a string (empty string if null or undefined).
          return (
            <div key={i} className={css.row}>
              <span className={css.key}>{key}</span>
              <span className={css.valueText}>
                {ValueComponent && value ? (
                  <ValueComponent value={value} {...valueComponentProps} />
                ) : (
                  `${value || ''}`
                )}
              </span>
            </div>
          );
        })}
      </div>
    </React.Fragment>
  );
};

export default DetailsTable;
