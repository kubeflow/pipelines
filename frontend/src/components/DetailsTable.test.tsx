/*
 * Copyright 2018 The Kubeflow Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

import DetailsTable from './DetailsTable';
import { render } from '@testing-library/react';

jest.mock('./Editor', () => {
  return ({ value }: { value: string }) => <pre data-testid='Editor'>{value}</pre>;
});

describe('DetailsTable', () => {
  it('shows no rows', () => {
    const { container } = render(<DetailsTable fields={[]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div />
      </div>
    `);
  });

  it('shows one row', () => {
    const { container } = render(<DetailsTable fields={[['key', 'value']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              value
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('shows a row with a title', () => {
    const { container } = render(<DetailsTable title='some title' fields={[['key', 'value']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div
          class="header2"
        >
          some title
        </div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              value
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('shows key and value for large values', () => {
    const { container } = render(
      <DetailsTable
        fields={[
          [
            'key',
            `Lorem Ipsum is simply dummy text of the printing and typesetting
      industry. Lorem Ipsum has been the industry's standard dummy text ever
      since the 1500s, when an unknown printer took a galley of type and
      scrambled it to make a type specimen book. It has survived not only five
      centuries, but also the leap into electronic typesetting, remaining
      essentially unchanged. It was popularised in the 1960s with the release
      of Letraset sheets containing Lorem Ipsum passages, and more recently
      with desktop publishing software like Aldus PageMaker including versions
      of Lorem Ipsum.`,
          ],
        ]}
      />,
    );
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry's standard dummy text ever
            since the 1500s, when an unknown printer took a galley of type and
            scrambled it to make a type specimen book. It has survived not only five
            centuries, but also the leap into electronic typesetting, remaining
            essentially unchanged. It was popularised in the 1960s with the release
            of Letraset sheets containing Lorem Ipsum passages, and more recently
            with desktop publishing software like Aldus PageMaker including versions
            of Lorem Ipsum.
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('shows key and value in row', () => {
    const { container } = render(<DetailsTable fields={[['key', 'value']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              value
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('shows key and JSON value in row', () => {
    const { container } = render(
      <DetailsTable fields={[['key', JSON.stringify([{ jsonKey: 'jsonValue' }])]]} />,
    );
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <pre
              data-testid="Editor"
            >
              [
        {
          "jsonKey": "jsonValue"
        }
      ]
            </pre>
          </div>
        </div>
      </div>
    `);
  });

  it('does render arrays as JSON', () => {
    const { container } = render(<DetailsTable fields={[['key', '[]']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <pre
              data-testid="Editor"
            >
              []
            </pre>
          </div>
        </div>
      </div>
    `);
  });

  it('does render empty object as JSON', () => {
    const { container } = render(<DetailsTable fields={[['key', '{}']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <pre
              data-testid="Editor"
            >
              {}
            </pre>
          </div>
        </div>
      </div>
    `);
  });

  it('does not render nulls as JSON', () => {
    const { container } = render(<DetailsTable fields={[['key', 'null']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              null
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('does not render numbers as JSON', () => {
    const { container } = render(<DetailsTable fields={[['key', '10']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              10
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('does not render strings as JSON', () => {
    const { container } = render(<DetailsTable fields={[['key', '"some string"']]} />);
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              "some string"
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('does not render booleans as JSON', () => {
    const { container } = render(
      <DetailsTable
        fields={[
          ['key1', 'true'],
          ['key2', 'false'],
        ]}
      />,
    );
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key1
            </span>
            <span
              class="valueText"
            >
              true
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key2
            </span>
            <span
              class="valueText"
            >
              false
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('shows keys and values for multiple rows', () => {
    const { container } = render(
      <DetailsTable
        fields={[
          ['key1', 'value1'],
          ['key2', JSON.stringify([{ jsonKey: 'jsonValue2' }])],
          ['key3', 'value3'],
          ['key4', 'value4'],
          ['key5', JSON.stringify({ jsonKey: { nestedJsonKey: 'jsonValue' } })],
          ['key6', 'value6'],
          ['key6', 'value7'],
          ['key', { key: 'foobar', bucket: 'bucket', endpoint: 's3.amazonaws.com' }],
        ]}
      />,
    );
    expect(container).toMatchInlineSnapshot(`
      <div>
        <div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key1
            </span>
            <span
              class="valueText"
            >
              value1
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key2
            </span>
            <pre
              data-testid="Editor"
            >
              [
        {
          "jsonKey": "jsonValue2"
        }
      ]
            </pre>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key3
            </span>
            <span
              class="valueText"
            >
              value3
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key4
            </span>
            <span
              class="valueText"
            >
              value4
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key5
            </span>
            <pre
              data-testid="Editor"
            >
              {
        "jsonKey": {
          "nestedJsonKey": "jsonValue"
        }
      }
            </pre>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key6
            </span>
            <span
              class="valueText"
            >
              value6
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key6
            </span>
            <span
              class="valueText"
            >
              value7
            </span>
          </div>
          <div
            class="row"
          >
            <span
              class="key"
            >
              key
            </span>
            <span
              class="valueText"
            >
              [object Object]
            </span>
          </div>
        </div>
      </div>
    `);
  });

  it('does render values with the provided valueComponent', () => {
    const ValueComponent: React.FC<any> = ({ value, ...rest }) => (
      <a data-testid='value-component' {...rest}>
        {JSON.stringify(value)}
      </a>
    );
    const { getByTestId } = render(
      <DetailsTable
        fields={[['key2', { key: 'foobar', bucket: 'bucket', endpoint: 's3.amazonaws.com' }]]}
        valueComponent={ValueComponent}
        valueComponentProps={{ extraprop: 'extra' }}
      />,
    );
    expect(getByTestId('value-component')).toMatchInlineSnapshot(`
      <a
        data-testid="value-component"
        extraprop="extra"
      >
        {"key":"foobar","bucket":"bucket","endpoint":"s3.amazonaws.com"}
      </a>
    `);
  });
});
