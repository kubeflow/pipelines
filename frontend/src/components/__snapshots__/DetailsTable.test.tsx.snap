// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsTable does not render booleans as JSON 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key1
      </span>
      <span
        className="valueText"
      >
        true
      </span>
    </div>
    <div
      className="row"
      key="1"
    >
      <span
        className="key"
      >
        key2
      </span>
      <span
        className="valueText"
      >
        false
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable does not render nulls as JSON 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        null
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable does not render numbers as JSON 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        10
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable does not render strings as JSON 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        "some string"
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable does render arrays as JSON 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <Editor
        cursorStart={1}
        editorProps={Object {}}
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="500px"
        highlightActiveLine={true}
        maxLines={20}
        minLines={3}
        mode="json"
        name="brace-editor"
        navigateToFileEnd={true}
        onChange={null}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={true}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={true}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="github"
        value="[]"
        width="100%"
        wrapEnabled={false}
      />
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable does render arrays as JSON 2`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <Editor
        cursorStart={1}
        editorProps={Object {}}
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="500px"
        highlightActiveLine={true}
        maxLines={20}
        minLines={3}
        mode="json"
        name="brace-editor"
        navigateToFileEnd={true}
        onChange={null}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={true}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={true}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="github"
        value="{}"
        width="100%"
        wrapEnabled={false}
      />
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows a row with a title 1`] = `
<Fragment>
  <div
    className="header"
  >
    some title
  </div>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        value
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows key and JSON value in row 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <Editor
        cursorStart={1}
        editorProps={Object {}}
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="500px"
        highlightActiveLine={true}
        maxLines={20}
        minLines={3}
        mode="json"
        name="brace-editor"
        navigateToFileEnd={true}
        onChange={null}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={true}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={true}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="github"
        value="[
  {
    \\"jsonKey\\": \\"jsonValue\\"
  }
]"
        width="100%"
        wrapEnabled={false}
      />
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows key and value for large values 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting
      industry. Lorem Ipsum has been the industry's standard dummy text ever
      since the 1500s, when an unknown printer took a galley of type and
      scrambled it to make a type specimen book. It has survived not only five
      centuries, but also the leap into electronic typesetting, remaining
      essentially unchanged. It was popularised in the 1960s with the release
      of Letraset sheets containing Lorem Ipsum passages, and more recently
      with desktop publishing software like Aldus PageMaker including versions
      of Lorem Ipsum.
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows key and value in row 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        value
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows keys and values for multiple rows 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key1
      </span>
      <span
        className="valueText"
      >
        value1
      </span>
    </div>
    <div
      className="row"
      key="1"
    >
      <span
        className="key"
      >
        key2
      </span>
      <Editor
        cursorStart={1}
        editorProps={Object {}}
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="500px"
        highlightActiveLine={true}
        maxLines={20}
        minLines={3}
        mode="json"
        name="brace-editor"
        navigateToFileEnd={true}
        onChange={null}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={true}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={true}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="github"
        value="[
  {
    \\"jsonKey\\": \\"jsonValue2\\"
  }
]"
        width="100%"
        wrapEnabled={false}
      />
    </div>
    <div
      className="row"
      key="2"
    >
      <span
        className="key"
      >
        key3
      </span>
      <span
        className="valueText"
      >
        value3
      </span>
    </div>
    <div
      className="row"
      key="3"
    >
      <span
        className="key"
      >
        key4
      </span>
      <span
        className="valueText"
      >
        value4
      </span>
    </div>
    <div
      className="row"
      key="4"
    >
      <span
        className="key"
      >
        key5
      </span>
      <Editor
        cursorStart={1}
        editorProps={Object {}}
        enableBasicAutocompletion={false}
        enableLiveAutocompletion={false}
        enableSnippets={false}
        focus={false}
        fontSize={12}
        height="500px"
        highlightActiveLine={true}
        maxLines={20}
        minLines={3}
        mode="json"
        name="brace-editor"
        navigateToFileEnd={true}
        onChange={null}
        onLoad={null}
        onPaste={null}
        onScroll={null}
        placeholder={null}
        readOnly={true}
        scrollMargin={
          Array [
            0,
            0,
            0,
            0,
          ]
        }
        setOptions={Object {}}
        showGutter={true}
        showPrintMargin={true}
        style={Object {}}
        tabSize={4}
        theme="github"
        value="{
  \\"jsonKey\\": {
    \\"nestedJsonKey\\": \\"jsonValue\\"
  }
}"
        width="100%"
        wrapEnabled={false}
      />
    </div>
    <div
      className="row"
      key="5"
    >
      <span
        className="key"
      >
        key6
      </span>
      <span
        className="valueText"
      >
        value6
      </span>
    </div>
    <div
      className="row"
      key="6"
    >
      <span
        className="key"
      >
        key6
      </span>
      <span
        className="valueText"
      >
        value7
      </span>
    </div>
  </div>
</Fragment>
`;

exports[`DetailsTable shows no rows 1`] = `
<Fragment>
  <div />
</Fragment>
`;

exports[`DetailsTable shows one row 1`] = `
<Fragment>
  <div>
    <div
      className="row"
      key="0"
    >
      <span
        className="key"
      >
        key
      </span>
      <span
        className="valueText"
      >
        value
      </span>
    </div>
  </div>
</Fragment>
`;
